src/AstTranslator.h

```diff
@@ -40,44 +48,266 @@ class TypeEnvironment;
  * and RAM structures.
  */
 class AstTranslator {
-public:
+private:
+    /** Map modified relation identifiers to original relation identifiers */
+    std::map<std::string, std::string> modifiedIdMap;
+
+    /** AST program */
+    const AstProgram* program;
+
+    /** Type environment */
+    const TypeEnvironment* typeEnv;
+
+private:
+    /**
+     * The location of some value in a loop nest.
+     */
+    struct Location {
+        int level;         // < the loop level
+        int component;     // < the component within the tuple created in the given level
+        std::string name;  // < name of the variable
+
+        bool operator==(const Location& loc) const {
+            return level == loc.level && component == loc.component;
+        }
+
+        bool operator!=(const Location& loc) const {
+            return !(*this == loc);
+        }
+
+        bool operator<(const Location& loc) const {
+            return level < loc.level || (level == loc.level && component < loc.component);
+        }
+
+        void print(std::ostream& out) const {
+            out << "(" << level << "," << component << ")";
+        }
+
+        friend std::ostream& operator<<(std::ostream& out, const Location& loc) {
+            loc.print(out);
+            return out;
+        }
+    };
+
     /**
-     * Converts the given relation identifier into a relation name.
+     * A class indexing the location of variables and record
+     * references within a loop nest resulting from the conversion
+     * of a rule.
      */
+    class ValueIndex {
+        /**
+         * The type mapping variables (referenced by their names) to the
+         * locations where they are used.
+         */
+        using variable_reference_map = std::map<std::string, std::set<Location>>;
+
+        /**
+         * The type mapping record init expressions to their definition points,
+         * hence the point where they get grounded/bound.
+         */
+        using record_definition_map = std::map<const AstRecordInit*, Location>;
+
+        /**
+         * The type mapping record init expressions to the loop level where
+         * they get unpacked.
+         */
+        using record_unpack_map = std::map<const AstRecordInit*, int>;
+
+        /**
+         * A map from AstAggregators to storage locations. Note, since in this case
+         * AstAggregators are indexed by their values (not their address) no standard
+         * map can be utilized.
+         */
+        using aggregator_location_map = std::vector<std::pair<const AstAggregator*, Location>>;
+
+        /** The index of variable accesses */
+        variable_reference_map var_references;
```

Comment:
Our variables follow a camelCase convention

Reply:
I have not touched this code; let's refactor AstTranslator after functors.